include/dlaf/eigensolver/tridiag_solver/rot.h

```diff
@@ -0,0 +1,243 @@
+    // Note:
+    // It would have been enough to just get the first tile from the beginning, and it would have
+    // worked anyway (thanks to the fact that panel has its own memorychunk and the first tile would
+    // keep alive the entire chunk, so also the part of other tiles, through its memory view).
+    // Anyway, it is more clean to get them all, and then just here use a single one to access the
+    // full column.
+    matrix::Tile<T, D>& tile_ws = all_ws[0];
+
+    // Note:
+    // The entire algorithm relies on a strong assumption about memory layout of all tiles involved,
+    // i.e. both the input tiles selected from the matrix and the tiles for the workspace.
+    // By relying on the fact that all of them are stored in a column-layout, we can work on column
+    // vectors ignoring their distribution over different tiles, and it would be enough just getting
+    // the pointer to the top head of the column and all other data can be easily accessed since it
+    // is stored contiguously after that.
+    // Ignoring tile organization of the memory comes handy also for communication. Indeed, during
+    // column exchange, just a single MPI operation per column is issued, instead of communicating
+    // independently the parts of column from each tile.
+    auto getColPtr = [&dist_sub, &tiles, &tile_ws](const SizeType col_index, const bool hasIt) -> T* {
+      if (hasIt) {
+        const LocalTileIndex tile_col{dist_sub.nextLocalTileFromGlobalElement<Coord::Row>(0),
+                                      dist_sub.nextLocalTileFromGlobalElement<Coord::Col>(col_index)};
+        const std::size_t linear_tile_col = to_sizet(dist_sub.localTileLinearIndex(tile_col));
+        return tiles[linear_tile_col].ptr(dist_sub.tileElementIndex({0, col_index}));
+      }
+      return tile_ws.ptr({0, 0});
+    };
+
+    // Change SoR for the rotations to just the range, by removing the offset
+    std::transform(rots.begin(), rots.end(), rots.begin(),
+                   [offset = i_begin * mb](const GivensRotation<T>& rot) {
+                     return GivensRotation<T>{rot.i - offset, rot.j - offset, rot.c, rot.s};
+                   });
+
+    const SizeType m = dist_sub.localSize().rows();
+
+    for (const GivensRotation<T>& rot : rots) {
+      const SizeType j_x = rot.i / mb;
+      const SizeType j_y = rot.j / mb;
+
+      const comm::IndexT_MPI rankColX = dist_sub.template rankGlobalTile<Coord::Col>(j_x);
+      const comm::IndexT_MPI rankColY = dist_sub.template rankGlobalTile<Coord::Col>(j_y);
+
+      const bool hasX = dist_sub.rankIndex().col() == rankColX;
+      const bool hasY = dist_sub.rankIndex().col() == rankColY;
+
+      if (!hasX && !hasY)
+        continue;
+
+      T* col_x = getColPtr(rot.i, hasX);
+      T* col_y = getColPtr(rot.j, hasY);
+
+      const bool hasBothXY = hasX && hasY;
+
+      std::vector<ex::unique_any_sender<>> cps;
+      if (!hasBothXY) {
+        // TODO compute TAG
```

Comment:
As said, my proposal is to get it from the caller, which could pass a tag unique for each call, ensuring the inter-call separation, that will be re-used by all intra-call communications that will happen.

This should™️ work correctly given current implementation details.

But, if in the future we will switch to a smarter/more advanced implementation where intra-call communications run in parallel, then the actual tag used would have to be a function of the current tag used. At that point the caller should ensure that all generated tags starting from the "seed" tag will be independent both intra and inter call...

@teonnik @rasolca what do you suggest to do about this tag?

Reply:
> But, if in the future we will switch to a smarter/more advanced implementation where intra-call communications run in parallel, then the actual tag used would have to be a function of the current tag used. At that point the caller should ensure that all generated tags starting from the "seed" tag will be independent both intra and inter call...

Yes, indeed. We will need to discuss how to manage tags across algorithms (if they use the same communicator) and across kernels of the same algorithm in case there is more than one kernel using `MPI_Isend` and `MPI_Irecv`. However, I think both of these questions are beyond the scope of this PR. The solution to have tag as part of the API so that a user can pass it to the function is IMO good.

I opened an issue about this : https://github.com/eth-cscs/DLA-Future/issues/700

Reply:
In the meanwhile, I did this https://github.com/eth-cscs/DLA-Future/pull/675/commits/a0d549ba75c2a9aac4e66a6ab4aaddfeb31f7a6d